[x86] Enable indirect tail calls with more arguments

zmodem opened the pull request:
X86ISelDAGToDAG's `isCalleeLoad` / `moveBelowOrigChain` tries to move the load instruction next to the call so they can be folded, but it would only allow a single CopyToReg node in between.

This patch makes it look through multiple CopyToReg, while being careful to only perform the transformation when the load+call can be folded.

Fixes #136848

zmodem committed:
(WORK IN PROGRESS) try to tail call address computation w/ more than two args

this folds more stuff, but also finds new breakages

Fixes #136848

zmodem committed:
fixes

rnk reviewed (commented):
Nice! The comments say to me this should be safe, but I'm honestly not sufficiently confident that I can read the SDAG logic well enough to know that it's correct. I would appreciate it if one of the other reviewers can take a look if they have time.

One thing you could do to build more confidence in the meantime would be to add more negative tests, like `load ; store ; call load`, perhaps substituting in other oddball memory clobbering write operations. Or just point at this this test if it exists.

zmodem committed:
format

zmodem committed:
don't allow TokenFactor with more than one use (found in chromium win64 build)

zmodem committed:
don't allow moving the load across inline asm

zmodem commented:
I also don't feel super confident with the DAG stuff. One positive is that we're good at asserting when there's an issue with the graph structure. That makes me more comfortable touching this code.

The current version builds Chromium for Linux (x64) and Windows (x86 and x64) asserts free (I found and fixed one issue).

The `load ; store ; call load` case is covered by https://github.com/llvm/llvm-project/blob/2f976956e5ccef566418853015fb6b31257aa69f/llvm/test/CodeGen/X86/fold-call.ll#L20-L26 Having inline asm in between also seems bad (a pre-existing bug?), so I added a check for that too.

The `hasOneUse()` checks I added generally prevent loops, which trigger asserts in the lit tests.

rnk commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -1353,6 +1376,22 @@ void X86DAGToDAGISel::PreprocessISelDAG() {
+        // the load address if folded into the tailcall. See how the
+        // X86tcret_6regs and X86tcret_1reg classes are used and defined.
+        unsigned NumRegs = 0;
+        for (unsigned I = 3, E = N->getNumOperands(); I != E; ++I) {
```
Is there a way to avoid the magic number 3?

zmodem replied:
I didn't find one, but I'll try to reduce the number of copies of this code.

rnk commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -1353,6 +1376,22 @@ void X86DAGToDAGISel::PreprocessISelDAG() {
+        // X86tcret_6regs and X86tcret_1reg classes are used and defined.
+        unsigned NumRegs = 0;
+        for (unsigned I = 3, E = N->getNumOperands(); I != E; ++I) {
+          if (isa<RegisterSDNode>(N->getOperand(I)))
```
We can skip XMM / FP register operands, so I would do a GPR64 class check here before counting up.

zmodem replied:
Done.

rnk commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -1353,6 +1376,22 @@ void X86DAGToDAGISel::PreprocessISelDAG() {
+        }
+        if (!Subtarget->is64Bit() && NumRegs > 1)
+          continue;
+        if (NumRegs > 6)
```
I think these values of 1 and 6 are informed by the SysV C calling conventions, and are incorrect for other calling conventions. You can probably construct a test case with custom conventions that use all available GPRs for parameters and starve the register allocator out.

I think this would fix a real issue on Windows x64, which by my reading only has 7 "volatile" GPRs:
https://learn.microsoft.com/en-us/cpp/build/x64-software-conventions?view=msvc-170#x64-register-usage
* RAX: return
* RCX, RDX, R8, R9: 4 param
* R10, R11: scratch

Total: 7

If there's a way to pass something in R11, maybe via the `nest` parameter, this might form a tail call we can't register allocate. Or maybe there's some other convention.

I think good general fixes would be to look at the CSR mask from the target calling convention and count up the available GPRs, subtract the number of GPR register operands from that total, and check that we have at least two available GPRs (for base + index).

Alternatively we could only do this fold for C calling conventions where it's known to be safe, if we convince ourselves that it's impossible to use R10 or R11 with the MS ABI convention.

zmodem replied:
You're right, Win64 has one less. I stole my code from the `X86tcret_6regs` fragment: https://github.com/llvm/llvm-project/blob/e58d227b09d533e2df644f827cedff8e206e0bfc/llvm/lib/Target/X86/X86InstrFragments.td#L676-L684

which is what the folding pattern for `TCRETURNmi64` uses: https://github.com/llvm/llvm-project/blob/e58d227b09d533e2df644f827cedff8e206e0bfc/llvm/lib/Target/X86/X86InstrCompiler.td#L1345-L1349

So that seems wrong for Win64.

I think the source of truth here is the register class which the folded instruction actually uses, which is `ptr_rc_tailcall` that gets defined by `X86RegisterInfo::getGPRsForTailCall`: https://github.com/llvm/llvm-project/blob/a2c1ff10eb930dd56be306dc0818d6ff31fff546/llvm/lib/Target/X86/X86RegisterInfo.cpp#L227-L239

That one seems to handle Win64 correctly, and also takes the calling convention into account in general.

---

So I think `X86tcret_6regs` should not hard-code 6, but check the `ptr_rc_tailcall` register class, and we should extract the code into a function that we can also use when moving the load.

And we should do the same for `X86tcret_1reg`, which is similar but has some differences: 

https://github.com/llvm/llvm-project/blob/e58d227b09d533e2df644f827cedff8e206e0bfc/llvm/lib/Target/X86/X86InstrFragments.td#L686-L699

It's checking whether the load uses a frame slot or a global, in which case it figures that doesn't use up any extra registers. I'm not 100% convinced that's true for the global case? And shouldn't we do the same check in `X86tcret_6regs`?

rnk reviewed (commented):
Thanks for the extra tests! Sorry I came up with more questions. My confidence in this code has increased, and I'd be OK to approve it after some iteration.

efriedma-quic commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -890,27 +890,50 @@ static bool isCalleeLoad(SDValue Callee, SDValue &Chain, bool HasCallSeq) {
+    // Moving across inline asm is not safe: it could do anything.
+    if (Chain.getNode()->getOpcode() == ISD::INLINEASM ||
+        Chain.getNode()->getOpcode() == ISD::INLINEASM_BR)
+      return false;
```
Please just allow specific nodes, and forbid anything unknown.  Trying to list out every possible relevant node is guaranteed to fall out of date at some point, even if you manage to come up with a complete list.

zmodem replied:
Done.

zmodem committed:
only allow moving the load across specific node types

zmodem committed:
only count GPRs, and compare against the size of getGPRsForTailCall()

zmodem committed:
extract the checks in X86tcret_6regs/1reg to a function and use it

zmodem committed:
format

zmodem commented:
Thanks for the review! This should probably be landed as separate patches, but let's keep it as one for the review.

I'd appreciate extra eyes on the code from `X86tcret_1reg` (see the XXX comment).

dpaoliello commented:
Ping... wondering what's happening with this patch: looks like it fixes an issue that I'm seeing where the load of a CFG dispatch isn't folded into the call.

efriedma-quic commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -3489,6 +3520,47 @@ static bool mayUseCarryFlag(X86::CondCode CC) {
+    const SDValue &BasePtr = Load->getBasePtr();
+    if (isa<FrameIndexSDNode>(BasePtr)) {
+      LoadGPRs = 0;
+    } else if (BasePtr->getNumOperands() &&
```
What's the opcode of BasePtr here?

efriedma-quic reviewed (commented):
I don't think anything is blocking this?  Added a minor comment from looking over it again.

dpaoliello commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -3489,6 +3520,47 @@ static bool mayUseCarryFlag(X86::CondCode CC) {
 
+bool X86DAGToDAGISel::checkTCRetRegUsage(SDNode *N, LoadSDNode *Load) const {
+  const X86RegisterInfo *RI = Subtarget->getRegisterInfo();
+  const TargetRegisterClass *TailCallGPRs = RI->getGPRsForTailCall(*MF);
```
Simplest thing would be something like:
```cpp
  const X86RegisterInfo *RI = Subtarget->getRegisterInfo();
  unsigned MaxGPRs;
  if (Subtarget->is64Bit()) {
    const TargetRegisterClass *TailCallGPRs = &X86::GR64_TCRegClass;
    assert(TailCallGPRs->contains(X86::RSP));
    assert(TailCallGPRs->contains(X86::RIP));
    MaxGPRs = TailCallGPRs->getNumRegs() - 2; // Can't use RSP or RIP for the address in general.
  } else {
    const TargetRegisterClass *TailCallGPRs = &X86::GR32_TCRegClass;
    assert(TailCallGPRs->contains(X86::ESP));
    MaxGPRs = TailCallGPRs->getNumRegs() - 1; // Can't use ESP for the address in general.
  }
```

@arsenm thoughts?

zmodem commented:
Sorry for dropping the ball here. I've kept the tab open with the intention of coming back and fixing this properly, but obviously that didn't happen yet.

The way I remember it, I kept finding more and more questionable things (like `X86tcret_6regs` and `X86tcret_1reg`), which I wanted to fix properly, and then of course the hard part becomes writing tests for it.

I still want to come back to this (maybe a project for the holidays :), but if someone wants to take over the PR and run with it, that's fine too.

antoniofrighetto commented on llvm/test/CodeGen/X86/fold-call-4.ll:
```diff
@@ -0,0 +1,26 @@
+; WIN: rex64 jmpq *(%rcx,%rdx,8)
+define void @f(ptr %table, i64 %idx, i64 %aux1, i64 %aux2, i64 %aux3) {
+entry:
+  %arrayidx = getelementptr inbounds ptr, ptr %table, i64 %idx
```
Should have a test addressing from a global too?

antoniofrighetto commented on llvm/test/CodeGen/X86/fold-call-4.ll:
```diff
@@ -0,0 +1,26 @@
+; RUN: llc < %s -mtriple=x86_64-unknown-linux-gnu | FileCheck %s --check-prefix=LIN
```
Should perhaps favour generating check lines via update_llc_test_checks for easier future maintenance?

zmodem replied:
This test has very few check lines. I think it's more clear this way.

zmodem committed:
Merge remote-tracking branch 'origin/main' into tail_call_with_two_args

zmodem committed:
rewrite and rename checkTCRetRegUsage to checkTCRetEnoughRegs

zmodem committed:
update tablegen

zmodem committed:
more tests

zmodem committed:
fixme

zmodem committed:
format

dpaoliello commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -875,27 +876,52 @@ static bool isCalleeLoad(SDValue Callee, SDValue &Chain, bool HasCallSeq) {
+
+    if (Chain.getOperand(0).getNode() == Callee.getNode())
+      return true;
+    if (Chain.getOperand(0).getOpcode() == ISD::TokenFactor &&
```
Can you please add a comment to explain what this is looking for?

zmodem committed:
add @test2_stack which tests the 'load from stack' case

zmodem committed:
Merge remote-tracking branch 'origin/main' into tail_call_with_two_args

efriedma-quic commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -3475,6 +3503,63 @@ static bool mayUseCarryFlag(X86::CondCode CC) {
+    // but it could apply to 64-bit too.
+    const SDValue &BasePtr = cast<LoadSDNode>(N->getOperand(1))->getBasePtr();
+    if (isa<FrameIndexSDNode>(BasePtr)) {
+      LoadGPRs -= 1; // Base is ESP, no reg needed.
```
Should this be `-= 2`?

efriedma-quic reviewed (approved):
LGTM with a few minor comments.

zmodem committed:
fix 'LoadGPRs -=' FrameIndexSDNode case

zmodem committed:
check for X86ISD::Wrapper explicitly

zmodem committed:
add comment

zmodem committed:
Merge remote-tracking branch 'origin/main' into tail_call_with_two_args

alexreinking commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -3479,6 +3507,65 @@ static bool mayUseCarryFlag(X86::CondCode CC) {
+      LoadGPRs -= 2; // Base is fixed index off ESP; no regs needed.
+    } else if (BasePtr.getOpcode() == X86ISD::Wrapper &&
+               isa<GlobalAddressSDNode>(BasePtr->getOperand(0))) {
+      assert(!getTargetMachine().isPositionIndependent());
```
Halide started tripping this assertion after merging this change. We were unconditionally requesting `Reloc::PIC_` when creating the `TargetMachine` for all targets, including x86-32-windows. This is the first time LLVM has asserted against that combination.

We can work around it by not requesting PIC on 32-bit Windows (it's meaningless for PE/COFF anyway). But I wonder if this is something LLVM should handle more gracefully... should `Reloc::PIC_` be rejected or silently ignored for Windows/COFF targets earlier in the pipeline (e.g. during `TargetMachine` creation) rather than propagating to instruction selection where it manifests as a confusing assertion?
